rgw: openssl 3

mdw-at-linuxbox opened the pull request:
[ this is an updated copy of PR #8339 ]
Augment configuration so it's easy to support http & https in one civetweb frontend.

Get rid of an incorrect warning when using openssl.

Document the thing so people can use it.

And load shared libraries by soname.

djzort commented:
please pull this asap! :)

mdw-at-linuxbox commented:
somebody rearranged *all* the cmake stuff.  I've force pushed a fix for that.

alimaredia commented:
@mdw-at-linuxbox the CMake aspect of this looks fine to me. @tchaikov anything look out of place to you?

yehudasa commented on src/rgw/rgw_main.cc:
```diff
@@ -403,10 +403,7 @@ int main(int argc, const char **argv)
-      config->get_val("port", 80, &port);
-
-      RGWProcessEnv env = { store, &rest, olog, port };
+      RGWProcessEnv env = { store, &rest, olog, 0 };
```
@mdw-at-linuxbox you can't remove the port here without addressing its use at RGWMongoose::init_env()

mdw-at-linuxbox replied:
Ok, that's related to : http://tracker.ceph.com/issues/16548 .  I sent mail 20160628 about that where I went through the code that does this.  The basic problem is that the current code path that uses that assumes that it can guess the server port based solely on whether the request came in with ssl or not.  That means not more than 1 ssl and 1 non-ssl port.  Do we really want to enforce that?

yehudasa replied:
@mdw-at-linuxbox we need to have some solution to that, a 0 port number is not a proper one.

theanalyst replied:
We could either 
1. parse port=".." the way civetweb does, ie. allow ip:port combo, r, s params that civetweb allows etc.

2.  alternatively, we can set override the env.port after initializing port from civetweb and getting the value of ports civetweb parses using the `mg_get_ports` api ?; ie something like this in rgw_civetweb_frontend:

```c++
(assume ports = new int [num ports] etc)
size_t len = mg_get_ports(ctx, num_ports, ports, ssl_ports)
if (len){
    env.port = ports[0] ;}
``` 
etc 
?

tchaikov commented on src/CMakeLists.txt:
```diff
@@ -818,13 +818,21 @@ if(${WITH_RADOSGW})
   target_include_directories(civetweb_common_objs PUBLIC
 	"${CMAKE_SOURCE_DIR}/src/civetweb/include")
-  if(HAVE_SSL)
+  if(USE_OPENSSL)
```
where is `USE_OPENSSL` initialized?

tchaikov commented on src/CMakeLists.txt:
```diff
@@ -818,13 +818,21 @@ if(${WITH_RADOSGW})
       "${SSL_INCLUDE_DIR}")
-  endif(HAVE_SSL)
+  endif(USE_OPENSSL)
+  if (DEFINED LIBSSL_SONAME)
```
could just put `if(LIBSSL_SONAME)`.

tchaikov commented on src/CMakeLists.txt:
```diff
@@ -818,13 +818,21 @@ if(${WITH_RADOSGW})
+  endif()
+
   add_subdirectory(rgw)
+
```
could remove this empty line.

smithfarm commented:
@mdw-at-linuxbox Does this PR fix http://tracker.ceph.com/issues/13718 ? If so, could you add

```
Fixes: http://tracker.ceph.com/issues/13718
```

immediately above the `Signed-off-by:` line of the main/most-relevant commit?

mdw-at-linuxbox commented:
Yes, this pull request would address #13718 - excepting that the problem Yehuda raised above makes this more complicated and probably a separate patch.

The top level CMakeLists.txt sets "HAVE_OPENSSL".  It should instead set USE_OPENSSL.

The "extra" blank line is because I had stuck a large glop of SSL obscurity in just before, and I didn't think the "add_subdirectory(rgw)" was obviously intentional without the extra blank lines.

Sure "if(LIBSSL_SONAME)" would work too - I take it we prefer this construction?

tchaikov commented:
> I take it we prefer this construction?

yeah. i think it's more concise.

@mdw-at-linuxbox ping

theanalyst commented:
this should probably address http://tracker.ceph.com/issues/16581 as well

mdw-at-linuxbox commented:
I've updated the branch wip-rgw-openssl-3.  I've removed "Replace + with , in civetweb port string.".  Whatever we do for that should be a separate pull request.  I've revised the doc accordingly.  I also spelled USE_OPENSSL right and removed DEFINED.  Finally I rebased on the latest master.  Passed my testing.  Hopefully gitbuilder will still like it.

theanalyst commented on src/rgw/rgw_main.cc:
```diff
@@ -403,10 +403,26 @@ int main(int argc, const char **argv)
+      if (portno <= 0 || portno > 65535) {
+	derr << "ERROR: port out of range " << port << dendl;
+	return -1;
+      }
```
what happens if we pass an ip:port combo which currently works, albeit setting SERVER_PORT to 0,  here don't we end up not starting rgw at all ?

mdw-at-linuxbox replied:
Well, actually, civetweb doesn't start up if you give it a bad port.  It fails in a much more confusing way later on.

However, you are right, it's possible to pass an ip:port in.  Or [ipv6]:port.  And append one of [sr,].  And it's possible to pass in non-zero bits >65535 in portno and have them ignored.  The ssl redir feature isn't useful to us currently, since we don't yet support listening for both ssl|non-ssl.  And we'll never pass in a , because the config file parser eats those elsewhere.  I don't mind breaking rgw for people who set high order bits in portno.

But being able to specify a specific ipaddr is useful, sigh.

theanalyst commented on CMakeLists.txt:
```diff
@@ -327,6 +327,47 @@ endif(WITH_XIO)
 
+if (${WITH_RADOSGW})
+  if (NOT DEFINED OPENSSL_FOUND)
+    message(STATUS "Looking for openssl anyays, because radosgw selected")
```
nit: typo, anyway

mdw-at-linuxbox commented:
I've force-pushed a new copy of this branch.  Rebased on latest master, there are now 4 commits:
a43ebf3 set -DUSE_IPV6 (NEW)
fe08886 document ssl
4470877 load by soname (FIX TYPO)
a43ebf3 parse and accept portno (UPDATED).

yehudasa commented on src/rgw/rgw_main.cc:
```diff
@@ -187,6 +187,30 @@ static void reloader_handler(int signum)
+static int parse_civetweb_portno(const string &port) {
+  char *portend;
+  const char *portstart = port.c_str();
+  const char *cp = strrchr(portstart, ':');
```
@mdw-at-linuxbox could just use port.find_last_of()

yehudasa commented on src/rgw/rgw_main.cc:
```diff
@@ -187,6 +187,30 @@ static void reloader_handler(int signum)
+  char *portend;
+  const char *portstart = port.c_str();
+  const char *cp = strrchr(portstart, ':');
+  if (cp)
```
@mdw-at-linuxbox in general we prefer explicit scoping

yehudasa commented on src/rgw/rgw_main.cc:
```diff
@@ -187,6 +187,30 @@ static void reloader_handler(int signum)
+    ++cp;
+  else
+    cp = portstart;
+  int portno = strtol(cp, &portend, 10);
```
@mdw-at-linuxbox use strict_strtol

yehudasa replied:
oh, nevermind, I see you need &portend

yehudasa commented on src/rgw/rgw_main.cc:
```diff
@@ -187,6 +187,30 @@ static void reloader_handler(int signum)
+  int portno = strtol(cp, &portend, 10);
+  if (cp == portend) {
+    derr << "ERROR: cannot parse port " << port << dendl;
+    return -1;
```
return -EINVAL

yehudasa replied:
also other occurances in this function

yehudasa commented on src/rgw/rgw_main.cc:
```diff
@@ -402,10 +426,15 @@ int main(int argc, const char **argv)
+      int portno;
+      config->get_val("port", "80", &port);
+      portno = parse_civetweb_portno(port);
+      if (portno == -1) {
```
if (portno < 0)

mdw-at-linuxbox commented:
port=0s can't work - you already pointed out why.  port=0 would result in a random port assignment.  How is that useful?

yehudasa commented:
@mdw-at-linuxbox see my comments on the last commit, other than that I think it's ok

djzort commented:
this seems to have gotten stuck?

mdw-at-linuxbox commented:
I've created a new PR that has just the soname fix and the documentation fix.  That is now PR # 11571.

liewegas commented:
need rebase?

mdw-at-linuxbox commented:
This PR is superseded by https://github.com/ceph/ceph/pull/11571 and https://github.com/ceph/ceph/pull/11776 so I'm closing this one.
